lib/fluent/plugin/buffer.rb

```diff
@@ -799,13 +803,23 @@ def write_step_by_step(metadata, data, format, splits_count, &block)
                   chunk.rollback
 
                   if split.size == 1 && original_bytesize == 0
-                    # It is obviously case that BufferChunkOverflowError should be raised here,
-                    # but if it raises here, already processed 'split' or
-                    # the proceeding 'split' will be lost completely.
-                    # so it is a last resort to delay raising such a exception
-                    errors << "concatenated/appended a #{adding_bytes} bytes record (nth: #{writing_splits_index}) is larger than buffer chunk limit size (#{@chunk_limit_size})"
-                    writing_splits_index += 1
-                    next
+                    # FIXME: adding_bytes should be implemented as an official chunk API in the future
+                    if adding_bytes.is_a?(Integer) && adding_bytes < @chunk_limit_size
```

Comment:
I'm not sure but `adding_bytes` might be able to calculate at caller side like this, because `adding_bytes` is cleared at `chunk.commit`, `chunk.rollback`, `chunk.purge`:

```diff
diff --git a/lib/fluent/plugin/buffer.rb b/lib/fluent/plugin/buffer.rb
index 5877af93..4df0dcb9 100644
--- a/lib/fluent/plugin/buffer.rb
+++ b/lib/fluent/plugin/buffer.rb
@@ -767,7 +767,7 @@ module Fluent
             raise ShouldRetry unless chunk.writable?
             staged_chunk_used = true if chunk.staged?
 
-            original_bytesize = chunk.bytesize
+            original_bytesize = before_bytesize = chunk.bytesize
             begin
               while writing_splits_index < splits.size
                 split = splits[writing_splits_index]
@@ -778,6 +778,7 @@ module Fluent
                     # so, keep already processed 'split' content here.
                     # (allow performance regression a bit)
                     chunk.commit
+                    before_bytesize = chunk.bytesize
                   else
                     big_record_size = formatted_split.bytesize
                     if big_record_size > @chunk_limit_size
@@ -788,6 +789,7 @@ module Fluent
                     elsif chunk.bytesize + big_record_size > @chunk_limit_size
                       # No doubt that the split is expected to cause size over, keep 'split' content here.
                       chunk.commit
+                      before_bytesize = chunk.bytesize
                     end
                   end
                 end
@@ -797,14 +799,14 @@ module Fluent
                 else
                   chunk.append(split, compress: @compress)
                 end
+                adding_bytes = chunk.bytesize - before_bytesize
 
                 if chunk_size_over?(chunk) # split size is larger than difference between size_full? and size_over?
-                  adding_bytes = chunk.instance_eval { @adding_bytes } || "N/A" # 3rd party might not have 'adding_bytes'
                   chunk.rollback
+                  before_bytesize = chunk.bytesize
 
                   if split.size == 1 && original_bytesize == 0
-                    # FIXME: adding_bytes should be implemented as an official chunk API in the future
-                    if adding_bytes.is_a?(Integer) && adding_bytes < @chunk_limit_size
+                    if adding_bytes < @chunk_limit_size
                       # already processed content is kept after rollback, then unstaged chunk should be queued.
                       # After that, re-process current split again (new chunk should be allocated, to do it, modify @stage and so on)
                       synchronize { @stage.delete(modified_metadata) }
```

Reply:
Thanks, this suggestion is more good approach to do the job.